lsp-pyls.el

```diff
@@ -353,53 +359,77 @@ parameters referenced in config."
+           ("pyls.plugins.pydocstyle.enabled" lsp-pyls-plugins-pydocstyle-enabled t)
+           ("pyls.plugins.pycodestyle.maxLineLength" lsp-pyls-plugins-pycodestyle-max-line-length)
+           ("pyls.plugins.pycodestyle.hangClosing" lsp-pyls-plugins-pycodestyle-hang-closing t)
+           ("pyls.plugins.pycodestyle.ignore" lsp-pyls-plugins-pycodestyle-ignore)
+           ("pyls.plugins.pycodestyle.select" lsp-pyls-plugins-pycodestyle-select)
+           ("pyls.plugins.pycodestyle.filename" lsp-pyls-plugins-pycodestyle-filename)
+           ("pyls.plugins.pycodestyle.exclude" lsp-pyls-plugins-pycodestyle-exclude)
+           ("pyls.plugins.pycodestyle.enabled" lsp-pyls-plugins-pycodestyle-enabled t)
+           ("pyls.plugins.pylint.enabled" lsp-pyls-plugins-pylint-enabled t)
+           ("pyls.plugins.pylint.args" lsp-pyls-plugins-pylint-args)
+           ("pyls.plugins.flake8.enabled" lsp-pyls-plugins-flake8-enabled)
+           ("pyls.plugins.flake8.exclude" lsp-pyls-plugins-flake8-exclude)
+           ("pyls.plugins.flake8.filename" lsp-pyls-plugins-flake8-filename)
+           ("pyls.plugins.flake8.hangClosing" lsp-pyls-plugins-flake8-hang-closing)
+           ("pyls.plugins.flake8.ignore" lsp-pyls-plugins-flake8-ignore)
+           ("pyls.plugins.flake8.maxLineLength" lsp-pyls-plugins-flake8-max-line-length)
+           ("pyls.plugins.flake8.select" lsp-pyls-plugins-flake8-select)
+           ("pyls.plugins.flake8.config" lsp-pyls-plugins-flake8-config)
+           ("pyls.plugins.preload.modules" lsp-pyls-plugins-preload-modules)
+           ("pyls.plugins.preload.enabled" lsp-pyls-plugins-preload-enabled t)
+           ("pyls.plugins.mccabe.threshold" lsp-pyls-plugins-mccabe-threshold)
+           ("pyls.plugins.mccabe.enabled" lsp-pyls-plugins-mccabe-enabled t)
+           ("pyls.plugins.jedi_symbols.all_scopes" lsp-pyls-plugins-jedi-symbols-all-scopes t)
+           ("pyls.plugins.jedi_symbols.enabled" lsp-pyls-plugins-jedi-symbols-enabled t)
+           ("pyls.plugins.jedi_signature_help.enabled" lsp-pyls-plugins-jedi-signature-help-enabled t)
+           ("pyls.plugins.jedi_references.enabled" lsp-pyls-plugins-jedi-references-enabled t)
+           ("pyls.plugins.jedi_hover.enabled" lsp-pyls-plugins-jedi-hover-enabled t)
+           ("pyls.plugins.jedi_definition.follow_builtin_imports" lsp-pyls-plugins-jedi-definition-follow-builtin-imports t)
+           ("pyls.plugins.jedi_definition.follow_imports" lsp-pyls-plugins-jedi-definition-follow-imports t)
+           ("pyls.plugins.jedi_definition.enabled" lsp-pyls-plugins-jedi-definition-enabled t)
+           ("pyls.plugins.jedi_completion.include_params" lsp-pyls-plugins-jedi-completion-include-params t)
+           ("pyls.plugins.jedi_completion.enabled" lsp-pyls-plugins-jedi-completion-enabled t)
+           ("pyls.configurationSources" lsp-pyls-configuration-sources))
+         (when lsp-pyls-plugins-jedi-use-pyenv-environment
+           (let ((pyenv-get-python-executable
```

Comment:
Do we need here a variable as well? I mean the logic might be - use the variable if the user has configured it if he hasn't try to automatically find the env. Also, it seems like the check for `lsp-pyls-plugins-jedi-use-pyenv-environment` should go in the lambda itself because you could have 2 projects - both of them have pyenv but you want to use pyenv for only one of them.

Reply:
Agree. With the variable you mentioned, added as `lsp-pyls-plugins-jedi-environment`, I am considering an approach that let users get the environment by `pyenv` and pass/set it for `lsp-mode`. However, since the custom-settings are for each `workspace` and users don't have information of `workspace`, they will have get it by `buffer-name`, `projectile-project-root` etc. As the `workspace` is defined inside `lsp-mode` and configured once, I prefer to calling `pyenv` on our side instead of on users side.

How do you think about it?

Reply:
Can you elaborate? I suggested the variable because the logic for automatic calculation might fail and in this case IMO it will be useful for the user to be able to set it manually. Also, the user might configure it using dir locals per project.

Reply:
I totally agree that users should be able to manually set the environment and added `lsp-pyls-plugins-jedi-environment` for this purpose. What I want to discuss is the "automatic" part, i.e. do we want to find the environment by `pyenv` in lsp-pyls or it should be the users responsiblity to do that and set the `lsp-pyls-plugins-jedi-environment` themselves. My concerns and opinions have been stated in the previous comment.

Reply:
I do not have enough python knowledge and I do not know what is the typical workflow is and what is the expected interaction `lsp-mode - pyenv - pyls`. I will let you and @mpanarin discuss it. From my point of view, the lsp-mode behaviour should be expected and as much stuff as possible should be working out of the box. For example, AFAIK pyls must be installed in the pyenv of the current project - I would like this to happen automatically - lsp-mode should detect that pyls is missing and offer to install it.

Reply:
> As the workspace is defined inside lsp-mode and configured once, I prefer to calling pyenv on our side instead of on users side.

I don't understand here what do you mean by "users side"?

Reply:
The approach I took would let stuff works out of the box. Let me discuss with @mpanarin since he have more context about how `plys` works. Thanks for your follow-on.